### PR TITLE
Add support for grants on new entity kinds

### PR DESCRIPTION
**Pull request by djsstarburst**
## Description

This PR privides a first-cut at syntax generalizing
privilege grants so that the entity kinds to which
the grants apply and the privileges for those grants
can be extended.
    
The PR also adds APIs to grant, deny, revoke and list grants
on arbitrary entity kinds.  Entity kinds TABLE and SCHEMA
are processed as before, but new entity kinds can be
supported by implementing the new methods on SystemAccessControl
and SystemSecurityMetadata.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 [hunk 1]
```diff
@@ -838,7 +838,17 @@ pathSpecification
...
 
 privilege
-    : CREATE | SELECT | DELETE | INSERT | UPDATE
+    : SELECT | UPDATE | INSERT | DELETE | CREATE
+    | LETTER+ (UNDERSCORE LETTER+)*
+    ;
+
+entityKind
+    : TABLE | SCHEMA
+    | LETTER+ (UNDERSCORE LETTER+)*
```
Comment (martint) [inline core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:847]: This uses lexer tokens that are defined as "fragments" and are not visible to the parser. I suspect that as it currently exists, this production never matches. Do you have any tests that involve an entity kind that's not table or schema?

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class GrantScope
+        extends Statement
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:26]: This should be a Node, not a Statement

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+{
+    private final Optional<String> entityKind;
+    private final QualifiedName name;
+
+    public GrantScope(Optional<String> entityKind, QualifiedName name)
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:31]: This constructor should not be needed. There's no scenario where the object should be instantiated without a location.

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        this(Optional.empty(), entityKind, name);
+    }
+
+    public GrantScope(NodeLocation location, Optional<String> entityKind, QualifiedName name)
+    {
+        this(Optional.of(location), entityKind, name);
+    }
+
+    public GrantScope(Optional<NodeLocation> location, Optional<String> entityKind, QualifiedName name)
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:41]: Same here.

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class GrantScope
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:25]: I'd call this `GrantObject`, as it's meant to identify the object for which privileges are being granted. FWIW, the SQL spec calls this construct an "object name", so the name I'm suggesting matches that more closely.

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 (hunk 1 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:847]: I fixed the fragment creating the appropriate productions.

HOWEVER, when I built a test in `TestSqlParser` for exotic privileges and entity kinds, I immediately ran into an ambiguity between `GRANT role TO grantee` and `GRANT funky_privilege ON t TO grantee`.  Before my changes, the ambiguity doesn't  exist, because the privileges, `SELECT`, `CREATE`, etc, are keywords and therefore don't collide with role names.

I don't know how to fix this ambiguity.  Do you have any suggestions?

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java (hunk 2 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java]: Changed.

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java (hunk 3 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java]: Removed.

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java (hunk 4 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java]: Removed.

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java (hunk 5 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java]: Changed.

File: core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4 (hunk 1 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4]: I refactored the grammar and all tests now pass.

**Comment by djsstarburst** [comment]
I believe this is in shape for review.  I refactored the grammar and all tests now pass.

File: core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4 [hunk 6]
```diff
@@ -119,35 +119,44 @@ statement
...
-        (IN catalog=identifier)?                                       #revokeRoles
-    | SET ROLE (ALL | NONE | role=identifier)
-        (IN catalog=identifier)?                                       #setRole
-    | GRANT
-        (privilege (',' privilege)* | ALL PRIVILEGES)
-        ON (SCHEMA | TABLE)? qualifiedName
-        TO grantee=principal
-        (WITH GRANT OPTION)?                                           #grant
+    | GRANT (
+        (
```
Comment (martint) [inline core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:123]: Have you tried using two separate labels for each of the branches instead of bundling them under `#genericGrant` ?

Comment (djsstarburst) [reply core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:123]: I did try.  They problem I saw an error in the IDE when was #grant inside the required outer parens.  AFAICT #tags are only allowed at the top level.

Comment (martint) [reply core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:123]: Then, I would try the following:

```
| GRANT privilegeOrRole (',' privilegeOrRole)*
            TO principal (',' principal)*
            (WITH ADMIN OPTION)?
            (GRANTED BY grantor)?
            (IN catalog=identifier)?   #grantRole
| GRANT ((privilegeOrRole (',' privilegeOrRole)*) | ALL PRIVILEGES)
            ON grantObject
            TO principal
            (WITH GRANT OPTION)? #grantPrivilege
```

Comment (djsstarburst) [reply core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4]: Awesome, @martint, that worked and all parser tests pass!

I pushed the improved version.

File: core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import static java.util.Objects.requireNonNull;
+
+public class EntityKindAndName
+{
+    private final String entityKind;
+    // TODO: Should this instead be List<String> like QualifiedName?
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java:24]: Yes, but it depends on what we're trying to achieve. Flattening a multipart QualifiedName into a string is problematic due to lack of escaping of special characters. For instance a name such as "a.b".c.d cannot be flattened to a string "a.b.c.d" without loss of information. If we want to be able to recover the catalog, schema and name, we need to keep them separate.

In fact. We may just want to separate catalog/schema/name as different elements.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 8]
```diff
@@ -1555,30 +1558,89 @@ public Node visitDropRole(SqlBaseParser.DropRoleContext context)
                 visitIfPresent(context.catalog, Identifier.class));
     }
 
+    @Override
+    public Node visitGrantPrivileges(SqlBaseParser.GrantPrivilegesContext context)
+    {
+        List<String> privileges;
+        if (context.ALL() != null) {
+            privileges = ImmutableList.of();
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:1566]: Why empty list? Just use `null` for this to remove any ambiguity about "empty means all" vs "empty means none"

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 9]
```diff
@@ -1555,30 +1558,89 @@ public Node visitDropRole(SqlBaseParser.DropRoleContext context)
...
+                privileges.isEmpty() ? Optional.empty() : Optional.of(privileges),
+                createGrantObject(getLocation(context), context.grantObject()),
+                getPrincipalSpecification(context.principal()),
+                context.OPTION() != null);
+    }
+
     @Override
     public Node visitGrantRoles(SqlBaseParser.GrantRolesContext context)
     {
+        // Check legality for a role grant.
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:1582]: Is this a statement about what this code is doing? If so, I don't see it. Or is it a "TO DO"?

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 10]
```diff
@@ -1555,30 +1558,89 @@ public Node visitDropRole(SqlBaseParser.DropRoleContext context)
...
                 ImmutableSet.copyOf(getPrincipalSpecifications(context.principal())),
                 context.OPTION() != null,
                 getGrantorSpecificationIfPresent(context.grantor()),
                 visitIfPresent(context.catalog, Identifier.class));
     }
 
+    @Override
+    public Node visitRevokePrivileges(SqlBaseParser.RevokePrivilegesContext context)
+    {
+        List<String> privileges = extractPrivileges(context.privilegeOrRole());
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:1595]: How is this method handling "ALL PRIVILEGES" ? Is empty list a side effect of the rule matching "ALL PRIVILEGES" instead of privilegeOrRole? If so, let's make the check explicit.

Also, see my comment above re: the semantics of empty list.

File: core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4 [hunk 11]
```diff
@@ -916,6 +916,16 @@ sqlStatementList
 
 privilege
     : CREATE | SELECT | DELETE | INSERT | UPDATE
+    | IDENTIFIER
```
[the remark this answers is not in the document]
Comment (djsstarburst) [reply core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4]: Privilege formatting changed.  I picked `IDENTIFIER` because I thought we would be unlikely to need or want quoted or digit identifiers.  But we likely want to allow `nonReserverd` tokens to be privileges, so `identifier` is the better choice.  Changed.

File: core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4 [hunk 12]
```diff
@@ -916,6 +916,16 @@ sqlStatementList
...
+    ;
+
+entityKind
+    : TABLE | SCHEMA
+    | IDENTIFIER
```
[the remark this answers is not in the document]
Comment (djsstarburst) [reply core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4]: Both changed.

File: core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+ * limitations under the License.
+ */
+package io.trino.spi.connector;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import static java.util.Objects.requireNonNull;
+
+public class EntityKindAndName
```
[the remark this answers is not in the document]
Comment (djsstarburst) [reply core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java]: LOL.  When I originally wrote the PR, Trino didn't yet allow records.

Changed to a record.

File: core/trino-spi/src/main/java/io/trino/spi/connector/EntityPrivilege.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+package io.trino.spi.connector;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import static java.util.Objects.requireNonNull;
+
+public class EntityPrivilege
```
[the remark this answers is not in the document]
Comment (djsstarburst) [reply core/trino-spi/src/main/java/io/trino/spi/connector/EntityPrivilege.java]: Changed to a record.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java [hunk 15]
```diff
@@ -3382,85 +3376,123 @@ public void testDeny()
...
+        // Failing cases
+        // TODO: Add more failing cases
+        assertThatThrownBy(() -> SQL_PARSER.createExpression("GRANT select, role23 TO role123"))
+                .isInstanceOf(ParsingException.class)
+                .hasMessageMatching("line 1:7: mismatched input 'select'. Expecting:.*");
+    }
+
+    private void testExoticPrivilegeGrants(String privilege, Optional<String> entityKind, PrincipalSpecification.Type type)
+    {
+        assertStatement(formatGrant("GRANT %s ON %st TO %su", privilege, entityKind, type),
```
[the remark this answers is not in the document]
Comment (djsstarburst) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:3419]: Unfortunately, `assertThat(statement(...))` requires accurate determination of the location of element, but these are tests whose SQL strings are constructed, so it is nontrivial to come up with the exact locations.

I just checked, and there are 352 uses of `assertStatement` and 81 uses of `assertThat(statement(...))`.  Unless you can think of a good way to come up with the locations, it seems better to me to allow `assertStatement` here.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java (hunk 8 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java]: Changed.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java (hunk 9 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java]: It was a leftover from the generic case.  Removed.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java (hunk 10 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java]: It was missing the required test for "ALL PRIVILEGES".  Added, and handled the empty list as suggested.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:3419]: We're trying to get rid of `assertStatement`. If we keep adding test cases that use it, we just make the migration harder for whoever ends up doing it. Why not do it the right way in the first place?

> requires accurate determination of the location of element, but these are tests whose SQL strings are constructed, so it is nontrivial to come up with the exact locations.

Then we should structure the tests in a different way so that we don't run into that problem. Let me take a closer look at what they are doing.

File: core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java (hunk 7 above)
Comment (djsstarburst) [reply core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java]: Let me know what you prefer, @martint, and I'll change it to conform.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (djsstarburst) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java]: I removed the generated tests, and wrote out the important tests long-hand using `assertThat(statement(...)).isEqualTo(...)`.

**Comment by djsstarburst** [comment]
The support for entity kind privileges was not handling ALL PRIVILEGES for exotic entity kinds.  I fixed that by adding a method on `Metadata` and `SystemSecurityMetadata` to get all privileges for an exotic entity kind.  That method is below.  I was conflicted about whether the new method should be added to `SystemAccessControl` or `SystemSecurityMetadata`.  In the end, I added it to `SystemSecurityMetadata`:

```
    /**
     * Gets all the EntityPrivileges associated with an entityKind.  Defines ALL PRIVILEGES
     * for the entityKind
     */
    default Set<EntityPrivilege> getAllEntityKindPrivileges(String entityKind)
    {
        throw new UnsupportedOperationException();
    }
```

File: core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java (hunk 7 above)
Comment (djsstarburst) [reply core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java]: I decided that `EntityKindAndName` should use `List<String>` as the name, and made that change.

I also added `SystemSecurityMetadata.getAllEntityKindPrivileges` to provide a way to know at parse time if a privilege is legal.

Finally, I created [this galaxy-trino branch](https://github.com/starburstdata/galaxy-trino/tree/djs/exotic-privileges) to support the new Trino entity privilege APIs once they are merged.

**Comment by mosabua** [comment]
I assume you are still working on this @djsstarburst ...

File: core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java [hunk 16]
```diff
@@ -44,6 +48,24 @@ public static Set<Privilege> parseStatementPrivileges(Node statement, Optional<L
...
 
+    public static Set<EntityPrivilege> parseEntityKindPrivileges(String entityKind, Metadata metadata, Optional<List<String>> privileges)
+    {
+        Set<EntityPrivilege> allPrivileges = metadata.getAllEntityKindPrivileges(entityKind);
+        if (privileges.isPresent()) {
+            return privileges.get().stream()
+                    .map(privilege -> {
+                        EntityPrivilege entityPrivilege = new EntityPrivilege(privilege.toUpperCase(Locale.ENGLISH));
+                        if (!allPrivileges.contains(entityPrivilege)) {
+                            throw new ParsingException("Privilege %s is not supported for entity kind %s".formatted(privilege, entityKind));
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java:59]: This should not be a ParsingException, but a TrinoException with INVALID_PRIVILEGE error code.

File: core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java [hunk 17]
```diff
@@ -44,6 +48,24 @@ public static Set<Privilege> parseStatementPrivileges(Node statement, Optional<L
         return privileges;
     }
 
+    public static Set<EntityPrivilege> parseEntityKindPrivileges(String entityKind, Metadata metadata, Optional<List<String>> privileges)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java:51]: This method is misnamed. There's no parsing involved here. I would rename it to "fetchEntityPrivileges" or similar.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 18]
```diff
@@ -1640,72 +1674,20 @@ public Node visitDeny(SqlBaseParser.DenyContext context)
...
-        if (context.qualifiedName() != null) {
-            tableName = Optional.of(getQualifiedName(context.qualifiedName()));
-        }
-
         return new ShowGrants(
                 getLocation(context),
-                context.TABLE() != null,
-                tableName);
+                context.grantObject() == null || context.grantObject().isEmpty() ? Optional.empty()
+                        : Optional.of(createGrantObject(getLocation(context), context.grantObject())));
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:1690]: Format as follows for readability (i.e., break the line at the `||` instead of at the `:` to make the visual layout more closely align with the evaluation precedence)

```suggestion
                context.grantObject() == null || 
                context.grantObject().isEmpty() ? Optional.empty() : Optional.of(createGrantObject(getLocation(context), context.grantObject())));
```

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java [hunk 19]
```diff
@@ -3113,7 +3095,7 @@ public Node visitGroupingOperation(SqlBaseParser.GroupingOperationContext contex
     }
 
     @Override
-    public Node visitUnquotedIdentifier(SqlBaseParser.UnquotedIdentifierContext context)
+    public Node visitUnquotedIdentifier(UnquotedIdentifierContext context)
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:3098]: Unrelated change?

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantObject.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+        return toStringHelper(this)
+                .add("entityKind", entityKind)
+                .add("qualifiedName", name)
+                .toString();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(getEntityKind(), getName());
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/tree/GrantObject.java:79]: Use the fields directly instead of going through the public getters, here and in equals(). While not the case today, it's conceivable that the getters might return a derived view of the internal state of the object that doesn't match the internal state.

File: core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public record EntityKindAndName(String entityKind, @JsonProperty List<String> name)
+{
+    public EntityKindAndName
+    {
+        requireNonNull(entityKind, "entityKind is null");
+        name = List.copyOf(requireNonNull(name, "name is null"));
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java:27]: requireNonNull not needed for this one. List.copyOf will fail if `name` is null.

**Review by martint — COMMENTED** [review]
Just a few minor comments, but otherwise looks good.

File: core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java (hunk 16 above)
Comment (djsstarburst) [reply core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java]: Changed.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java (hunk 18 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java]: Changed.

File: core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java (hunk 19 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java]: Reverted.

File: core/trino-parser/src/main/java/io/trino/sql/tree/GrantObject.java (hunk 20 above)
Comment (djsstarburst) [reply core/trino-parser/src/main/java/io/trino/sql/tree/GrantObject.java]: Good point, and a lesson I'll remember in the future.  Changed.

File: core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java (hunk 21 above)
Comment (djsstarburst) [reply core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java]: Removed.

**Comment by djsstarburst** [comment]
Thanks very much for reviewing, @martint!  I made the changes you suggested.

**Comment by mosabua** [comment]
💥  @djsstarburst !

**Comment by djsstarburst** [comment]
Yay, it's merged!
